Development/nmos/constraints.cpp

```diff
@@ -0,0 +1,284 @@
+            {
+                return false;
+            }
+            // subconstraint should have minimum or enum if constraint has minimum
+            if (constraint.has_field(nmos::fields::constraint_minimum) && !subconstraint.has_field(nmos::fields::constraint_enum) && !subconstraint.has_field(nmos::fields::constraint_minimum))
+            {
+                return false;
+            }
+            // subconstraint should have maximum or enum if constraint has maximum
+            if (constraint.has_field(nmos::fields::constraint_maximum) && !subconstraint.has_field(nmos::fields::constraint_enum) && !subconstraint.has_field(nmos::fields::constraint_maximum))
+            {
+                return false;
+            }
+            if (constraint.has_field(nmos::fields::constraint_minimum) && subconstraint.has_field(nmos::fields::constraint_minimum))
+            {
+                const auto& constraint_min = nmos::fields::constraint_minimum(constraint);
+                const auto& subconstraint_min = nmos::fields::constraint_minimum(subconstraint);
+
+                if (constraint_min != subconstraint_min)
+                {
+                    if (!constraint_comp(constraint_min, subconstraint_min))
+                    {
+                        return false;
+                    }
+                }
+            }
+            if (constraint.has_field(nmos::fields::constraint_maximum) && subconstraint.has_field(nmos::fields::constraint_maximum))
+            {
+                const auto& constraint_max = nmos::fields::constraint_maximum(constraint);
+                const auto& subconstraint_max = nmos::fields::constraint_maximum(subconstraint);
+
+                if (constraint_comp(constraint_max, subconstraint_max))
+                {
+                    return false;
+                }
+            }
+            // subconstraint enum values should match constraint
+            if (subconstraint.has_field(nmos::fields::constraint_enum))
+            {
+                const auto& subconstraint_enum_values = nmos::fields::constraint_enum(subconstraint).as_array();
+                return subconstraint_enum_values.end() == std::find_if_not(subconstraint_enum_values.begin(), subconstraint_enum_values.end(), [&constraint](const web::json::value& enum_value)
+                {
+                    return match_constraint(enum_value, constraint);
+                });
+            }
+            return true;
+        }
+
+        // Constraint Set B is a subset of Constraint Set A if all Parameter Constraints of Constraint Set A are present in Constraint Set B, and for each Parameter Constraint
+        // that is present in both, the Parameter Constraint of Constraint Set B is a subconstraint of the Parameter Constraint of Constraint Set A.
+        bool is_constraint_subset(const web::json::value& constraint_set, const web::json::value& constraint_subset, bool merge_left_to_right)
```

Comment:
I'm still not seeing why `merge` parameter is necessary to determine if B is a subconstraint of A - if `constraint_set` (A) has any parameter constraints not in `constraint_subset`(B), then B is not a constraint subset, whereas it doesn't effect the result if B has any parameter constraints not in A.